File: scripts/generate-ciqual.py
#!/usr/bin/env python3
"""
Generate assets/data/ciqual.json from CIQUAL 2025 XML files.

Usage:
    python3 scripts/generate-ciqual.py

Downloads alim.xml (~1.5 MB) and compo.xml (~69 MB) from data.gouv.fr,
then generates a trimmed JSON with only the 6 nutritional values needed.
Temporary XML files are cached in /tmp to avoid re-downloading.
"""

import json
import os
import sys
import urllib.request
from xml.etree.ElementTree import iterparse

ALIM_URL = (
    "https://entrepot.recherche.data.gouv.fr/api/access/datafile/"
    ":persistentId?persistentId=doi:10.57745/OH8KXC"
)
COMPO_URL = (
    "https://entrepot.recherche.data.gouv.fr/api/access/datafile/"
    ":persistentId?persistentId=doi:10.57745/O73GDX"
)

CONST_CODES = {
    "328":   "kcal",
    "25000": "proteines",
    "31000": "glucides",
    "40000": "lipides",
    "34100": "fibres",
    "10004": "sel",
}

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
OUTPUT_PATH = os.path.join(SCRIPT_DIR, "..", "assets", "data", "ciqual.json")
TMP_ALIM = "/tmp/ciqual_alim.xml"
TMP_COMPO = "/tmp/ciqual_compo.xml"


def parse_teneur(value: str) -> float | None:
    v = value.strip()
    if not v or v == "-":
        return None
    if v.lower() in ("traces", "trace"):
        return 0.0
    if v.startswith("<"):
        return 0.0
    try:
        return float(v.replace(",", "."))
    except ValueError:
        return None


def download(url: str, dest: str) -> None:
    print(f"  Downloading {os.path.basename(dest)}...", flush=True)
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req) as resp, open(dest, "wb") as f:
        downloaded = 0
        while chunk := resp.read(65536):
            f.write(chunk)
            downloaded += len(chunk)
            print(f"\r  {downloaded // 1024 // 1024} MB", end="", flush=True)
    print(f"\r  {os.path.getsize(dest) // 1024} KB — done        ")


def parse_alim(path: str) -> dict[str, str]:
    foods: dict[str, str] = {}
    for _, elem in iterparse(path, events=("end",)):
        if elem.tag == "alim":
            code = elem.get("alim_code", "").strip()
            name = elem.get("alim_nom_fr", "").strip()
            if code and name:
                foods[code] = name
            elem.clear()
    return foods


def parse_compo(path: str) -> dict[str, dict[str, float]]:
    data: dict[str, dict[str, float]] = {}
    target = set(CONST_CODES.keys())
    count = 0
    for _, elem in iterparse(path, events=("end",)):
        if elem.tag == "compo":
            const_code = elem.get("const_code", "")
            if const_code in target:
                alim_code = elem.get("alim_code", "")
                val = parse_teneur(elem.get("teneur", "-"))
                if val is not None:
                    data.setdefault(alim_code, {})[CONST_CODES[const_code]] = round(val, 1)
                count += 1
                if count % 10000 == 0:
                    print(f"\r  {count} entries parsed...", end="", flush=True)
            elem.clear()
    print(f"\r  {count} relevant entries parsed          ")
    return data


def main() -> None:
    print("=== CIQUAL 2025 JSON generator ===\n")
    os.makedirs(os.path.dirname(OUTPUT_PATH), exist_ok=True)

    print("1/4 — Aliments XML")
    if os.path.exists(TMP_ALIM):
        print(f"  Using cached {TMP_ALIM}")
    else:
        download(ALIM_URL, TMP_ALIM)

    print("2/4 — Compositions XML (large file, ~2 min)")
    if os.path.exists(TMP_COMPO):
        print(f"  Using cached {TMP_COMPO}")
    else:
        download(COMPO_URL, TMP_COMPO)

    print("3/4 — Parsing aliments...")
    foods = parse_alim(TMP_ALIM)
    print(f"  {len(foods)} foods found")

    print("4/4 — Parsing compositions...")
    compo = parse_compo(TMP_COMPO)
    print(f"  {len(compo)} compositions found")

    result = []
    for alim_code, name in foods.items():
        nutr = compo.get(alim_code, {})
        kcal = nutr.get("kcal")
        if kcal is None:
            continue
        result.append({
            "id": alim_code,
            "nom": name,
            "kcal": kcal,
            "proteines": nutr.get("proteines", 0.0),
            "glucides": nutr.get("glucides", 0.0),
            "lipides": nutr.get("lipides", 0.0),
            "fibres": nutr.get("fibres"),
            "sel": nutr.get("sel"),
        })

    result.sort(key=lambda x: x["nom"])

    with open(OUTPUT_PATH, "w", encoding="utf-8") as f:
        json.dump(result, f, ensure_ascii=False, separators=(",", ":"))

    size_kb = os.path.getsize(OUTPUT_PATH) // 1024
    print(f"\n✓ {OUTPUT_PATH}")
    print(f"  {len(result)} foods — {size_kb} KB")
    print("\nRestart Expo after regenerating to pick up the new file.")


if __name__ == "__main__":
    main()

File: scripts/generate-ciqual.mjs
#!/usr/bin/env node
/**
 * Generate assets/data/ciqual.json from CIQUAL 2025 XML files.
 * Uses only Node.js built-in modules — no npm install needed.
 *
 * Usage: node scripts/generate-ciqual.mjs
 */

import { createWriteStream, createReadStream, statSync, existsSync, mkdirSync, writeFileSync } from 'node:fs'
import https from 'node:https'
import http from 'node:http'
import { createInterface } from 'node:readline'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = dirname(fileURLToPath(import.meta.url))

const ALIM_URL =
  'https://entrepot.recherche.data.gouv.fr/api/access/datafile/' +
  ':persistentId?persistentId=doi:10.57745/OH8KXC'
const COMPO_URL =
  'https://entrepot.recherche.data.gouv.fr/api/access/datafile/' +
  ':persistentId?persistentId=doi:10.57745/O73GDX'

// Codes constituants CIQUAL 2025
const CONST_CODES = {
  '328':   'kcal',
  '25000': 'proteines',
  '31000': 'glucides',
  '40000': 'lipides',
  '34100': 'fibres',
  '10004': 'sel',
}

const OUTPUT_PATH = join(__dirname, '..', 'assets', 'data', 'ciqual.json')
const TMP_ALIM  = '/tmp/ciqual_alim.xml'
const TMP_COMPO = '/tmp/ciqual_compo.xml'

// ── helpers ────────────────────────────────────────────────────────────────

function parseTeneur(v = '-') {
  const s = (v || '').trim()
  if (!s || s === '-') return null
  if (s.toLowerCase().startsWith('trace')) return 0
  if (s.startsWith('<')) return 0
  const n = parseFloat(s.replace(',', '.'))
  return isNaN(n) ? null : n
}

/** Extract text content from <tagName>value</tagName> on a single line */
function extractTag(line, tagName) {
  const re = new RegExp(`<${tagName}(?:\\s[^>]*)?>([^<]*)<\\/${tagName}>`)
  const m = re.exec(line)
  return m ? m[1].trim() : null
}

function download(url, dest) {
  return new Promise((resolve, reject) => {
    function request(currentUrl, hops = 0) {
      if (hops > 8) return reject(new Error('Too many redirects'))
      const mod = currentUrl.startsWith('https') ? https : http
      mod.get(currentUrl, { headers: { 'User-Agent': 'Mozilla/5.0' } }, (res) => {
        if ([301, 302, 303].includes(res.statusCode)) {
          res.resume()
          return request(res.headers.location, hops + 1)
        }
        if (res.statusCode !== 200) {
          res.resume()
          return reject(new Error(`HTTP ${res.statusCode}`))
        }
        const file = createWriteStream(dest)
        let bytes = 0
        res.on('data', (chunk) => {
          bytes += chunk.length
          process.stdout.write(`\r  ${(bytes / 1024 / 1024).toFixed(1)} MB...`)
        })
        res.pipe(file)
        file.on('finish', () => {
          file.close()
          const kb = Math.round(statSync(dest).size / 1024)
          console.log(`\r  ${kb} KB — done            `)
          resolve()
        })
        file.on('error', reject)
      }).on('error', reject)
    }
    request(url)
  })
}

/**
 * XML format:
 *   <ALIM>
 *     <alim_code> 1000 </alim_code>
 *     <alim_nom_fr> Pastis </alim_nom_fr>
 *     ...
 *   </ALIM>
 */
async function parseAlim(path) {
  const foods = {}
  let inBlock = false
  let code = null
  let nom = null

  const rl = createInterface({ input: createReadStream(path), crlfDelay: Infinity })
  for await (const line of rl) {
    const t = line.trim()
    if (t === '<ALIM>') { inBlock = true; code = null; nom = null; continue }
    if (t === '</ALIM>') {
      if (code && nom) foods[code] = nom
      inBlock = false; continue
    }
    if (!inBlock) continue
    if (!code) code = extractTag(t, 'alim_code')
    if (!nom)  nom  = extractTag(t, 'alim_nom_fr')
  }
  return foods
}

/**
 * XML format:
 *   <COMPO>
 *     <alim_code> 1000 </alim_code>
 *     <const_code> 328 </const_code>
 *     <teneur> 274 </teneur>
 *     ...
 *   </COMPO>
 */
async function parseCompo(path) {
  const data = {}
  const target = new Set(Object.keys(CONST_CODES))
  let inBlock = false
  let alimCode = null
  let constCode = null
  let teneur = null
  let count = 0

  const rl = createInterface({ input: createReadStream(path), crlfDelay: Infinity })
  for await (const line of rl) {
    const t = line.trim()
    if (t === '<COMPO>') { inBlock = true; alimCode = null; constCode = null; teneur = null; continue }
    if (t === '</COMPO>') {
      if (alimCode && constCode && target.has(constCode)) {
        const val = parseTeneur(teneur)
        if (val !== null) {
          data[alimCode] ??= {}
          data[alimCode][CONST_CODES[constCode]] = Math.round(val * 10) / 10
        }
        if (++count % 10000 === 0) process.stdout.write(`\r  ${count} entries...`)
      }
      inBlock = false; continue
    }
    if (!inBlock) continue
    if (!alimCode)  alimCode  = extractTag(t, 'alim_code')
    if (!constCode) constCode = extractTag(t, 'const_code')
    if (!teneur)    teneur    = extractTag(t, 'teneur')
  }
  console.log(`\r  ${count} relevant entries parsed       `)
  return data
}

// ── main ────────────────────────────────────────────────────────────────────

console.log('=== CIQUAL 2025 JSON generator ===\n')
mkdirSync(dirname(OUTPUT_PATH), { recursive: true })

console.log('1/4 — alim.xml')
if (existsSync(TMP_ALIM)) console.log(`  Using cached ${TMP_ALIM}`)
else await download(ALIM_URL, TMP_ALIM)

console.log('2/4 — compo.xml (~67 MB, quelques minutes)')
if (existsSync(TMP_COMPO)) console.log(`  Using cached ${TMP_COMPO}`)
else await download(COMPO_URL, TMP_COMPO)

console.log('3/4 — Parsing aliments...')
const foods = await parseAlim(TMP_ALIM)
console.log(`  ${Object.keys(foods).length} aliments trouvés`)

console.log('4/4 — Parsing compositions...')
const compo = await parseCompo(TMP_COMPO)

const result = Object.entries(foods)
  .reduce((acc, [id, nom]) => {
    const n = compo[id] ?? {}
    if (n.kcal == null) return acc
    acc.push({
      id,
      nom,
      kcal:      n.kcal,
      proteines: n.proteines ?? 0,
      glucides:  n.glucides  ?? 0,
      lipides:   n.lipides   ?? 0,
      fibres:    n.fibres    ?? null,
      sel:       n.sel       ?? null,
    })
    return acc
  }, [])
  .sort((a, b) => a.nom.localeCompare(b.nom, 'fr'))

writeFileSync(OUTPUT_PATH, JSON.stringify(result), 'utf8')
const kb = Math.round(statSync(OUTPUT_PATH).size / 1024)
console.log(`\n✓ ${OUTPUT_PATH}`)
console.log(`  ${result.length} aliments — ${kb} KB`)
console.log('\nRedémarre Expo pour prendre en compte le fichier.')
